fd_linux.go

```diff
@@ -1,45 +1,213 @@
+	SetNonblocking(name string) error
 }
 
-var _ fd = &sysFD{}
+var _ listenFD = &sysListenFD{}
 
-// sysFD is the system call implementation of fd.
-type sysFD struct {
-	fd int
+// A sysListenFD is the system call implementation of listenFD.
+type sysListenFD struct {
+	// These fields should never be non-zero at the same time.
+	fd int      // Used in blocking mode.
+	f  *os.File // Used in non-blocking mode.
 }
 
-func (fd *sysFD) Accept4(flags int) (fd, unix.Sockaddr, error) {
-	// Returns a regular file descriptor, must be wrapped in another
-	// sysFD for it to work properly.
-	nfd, sa, err := unix.Accept4(fd.fd, flags)
+// newListenFD creates a sysListenFD in its default blocking mode.
+func newListenFD() (*sysListenFD, error) {
+	fd, err := unix.Socket(unix.AF_VSOCK, unix.SOCK_STREAM, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sysListenFD{
+		fd: fd,
+	}, nil
+}
+
+// Blocking mode methods.
+
+func (lfd *sysListenFD) Bind(sa unix.Sockaddr) error         { return unix.Bind(lfd.fd, sa) }
+func (lfd *sysListenFD) Getsockname() (unix.Sockaddr, error) { return unix.Getsockname(lfd.fd) }
+func (lfd *sysListenFD) Listen(n int) error                  { return unix.Listen(lfd.fd, n) }
+
+func (lfd *sysListenFD) SetNonblocking(name string) error {
+	// From now on, we must perform non-blocking I/O, so that our
+	// net.Listener.Accept method can be interrupted by closing the socket.
+	if err := unix.SetNonblock(lfd.fd, true); err != nil {
+		return err
+	}
+
+	// Transition from blocking mode to non-blocking mode.
+	lfd.f = os.NewFile(uintptr(lfd.fd), name)
+
+	return nil
+}
+
+// EarlyClose is a blocking version of Close, only used for cleanup before
+// entering non-blocking mode.
+func (lfd *sysListenFD) EarlyClose() error { return unix.Close(lfd.fd) }
+
+// Non-blocking mode methods.
+
+func (lfd *sysListenFD) Accept4(flags int) (connFD, unix.Sockaddr, error) {
+	rc, err := lfd.f.SyscallConn()
 	if err != nil {
 		return nil, nil, err
 	}
 
-	return &sysFD{fd: nfd}, sa, nil
+	var (
+		nfd int
+		sa  unix.Sockaddr
+	)
+
+	doErr := rc.Read(func(fd uintptr) bool {
+		nfd, sa, err = unix.Accept4(int(fd), flags)
+
+		// When the socket is in non-blocking mode, we might see
+		// EAGAIN and end up here. In that case, return false to
+		// let the poller wait for readiness. See the source code
+		// for internal/poll.FD.RawRead for more details.
+		return err != unix.EAGAIN
+	})
+	if doErr != nil {
+		return nil, nil, doErr
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create a non-blocking connFD which will be used to implement net.Conn.
+	cfd := &sysConnFD{fd: nfd}
+	return cfd, sa, nil
+}
+
+func (lfd *sysListenFD) Close() error {
+	var err error
+	doErr := fdcontrol(lfd.f, func(fd int) {
+		err = unix.Close(fd)
+	})
+	if doErr != nil {
+		return doErr
+	}
+
+	// We must also close the runtime network poller file descriptor for
+	// net.Listener.Accept to stop blocking.
+	_ = lfd.f.Close()
```

Comment:
I have some concerns about this (also wrt the comment above). The scenario I am about to describe may in fact be a non-issue, but I'm not sure it is, based on my reading of `internal/poll` internals.

Imagine we are `fd=5`. If we are here, we have already called `unix.Close(5)` in the `fdcontrol` block above. `fd=5` is closed. But `lfd.f` still thinks it is (and owns) `fd=5`!

Between the time when we called `unix.Close` above and this point, in a completely unrelated part of the program, a different goroutine does `os.Open("something")`, and the result is assigned `fd=5` by the kernel, since `fd=5` is already closed. We promptly proceed to do `ldf.f.Close()`, closing `fd=5` and sweeping the file descriptor from under the unrelated client.

Reply:
Thank you for your analysis. I have updated this.